analyzers/src/SonarAnalyzer.CSharp/Rules/ImplementIDisposableCorrectly.cs

```diff
@@ -269,15 +268,20 @@ private void VerifyDispose(MethodDeclarationSyntax disposeMethod, bool isSealedC
                 typeSymbol.GetMembers()
                           .OfType<IMethodSymbol>()
                           .Where(predicate)
-                          .SelectMany(m => m.DeclaringSyntaxReferences)
-                          .Select(r => r.GetSyntax());
-
-            private static bool HasVirtualDisposeBool(INamedTypeSymbol symbol) =>
-                symbol.GetSelfAndBaseTypes()
-                      .SelectMany(t => t.GetMembers())
-                      .OfType<IMethodSymbol>()
-                      .Where(IsDisposeBool)
-                      .Any(methodSym => !methodSym.IsAbstract);
+                          .SelectMany(symbol => symbol.PartialImplementationPart?.DeclaringSyntaxReferences ?? symbol.DeclaringSyntaxReferences)
+                          .Select(reference => reference.GetSyntax());
+
+            private static IEnumerable<SyntaxNode> FindMethodImplementation(INamedTypeSymbol typeSymbol, Func<IMethodSymbol, bool> predicate, TypeDeclarationSyntax typeDeclarationSyntax) =>
+                FindMethodDeclarations(typeSymbol, predicate)
+                    .OfType<BaseMethodDeclarationSyntax>()
+                    .Where(syntax => typeDeclarationSyntax.Contains(syntax) && (syntax.HasBodyOrExpressionBody() || syntax.Modifiers.AnyOfKind(SyntaxKind.AbstractKeyword)));
```

Comment:
```suggestion
                    .Where(x => typeDeclarationSyntax.Contains(x) && (x.HasBodyOrExpressionBody() || x.Modifiers.AnyOfKind(SyntaxKind.AbstractKeyword)));
```

(it's clear that it's a BaseMethodDeclarationSyntax here)

Reply:
why do you check for `AbstractKeyword` here? I couldn't find it in the tests

Reply:
We check for `AbstractKeyword` since I want to preserve the existing functionality for abstract methods and raise secondary issues for them. 

You can find the test cases with abstract methods here: https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ImplementIDisposableCorrectly.AbstractClass.cs#L14